src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g

```diff
@@ -1966,6 +2014,17 @@ IDENT
             if (mTreatEnumAsKeyword && "enum".equals($getText)) {
                 $setType(ENUM);
             }
+            // records must not be in literal list, else legacy usage of records
+            // i.e. "LogRecord record = log();" as a variable and method name
+            // cause false positives. The 'keyword as identifier' problem.
+            if (mTreatRecordAsKeyword
+                && "record".equals($getText)
+                && LA(2) != '=' && LA(1) != ';'
+                && LA(1) != ')' && LA(1) != '.'
+                && LA(1) != ',' && LA(1) != '('
+                && LA(2) != ':') {
+                $setType(RECORD);
```

Comment:
I wondered about an approach like this as well -- I think it may be hard to catch all the cases, though (e.g. `record[]`, `record ->`, `record<T>`, etc.), or to really decide keyword vs. identifier in the lexer (what about `class record extends Object`), and seems like it could be difficult to maintain.

It makes me wonder again about leaving `record` as an `IDENT` as far as the lexer is concerned, and just trying to deal with it in the grammar rule. I know you have already been down multiple paths with predicates -- I will take a look myself a little later tonight and see if anything strikes me...

Reply:
[This](https://github.com/nmancus1/checkstyle/commit/7c9102b62960fa1f1b3790251aa8bc9045aeaafc) actually works pretty great for records, but messes up quite a few other things.  I got the idea from [here](https://www.google.com/url?q=https%3A%2F%2Fgithub.com%2Fjszheng%2Fpy3antlr4book%2Fblob%2Fmaster%2F12-id_as_keyword%2FPredKeyword.g4&sa=D&sntz=1&usg=AFQjCNFN0TC9Wyh_WM_05f3ZSfTWJu6tYQ).